### ISSUE TITLE
Masking allowed actions at root node

### ISSUE THREAD

**Issue by fdrake76** [werner-duvaud/muzero-general#5]
Hi Werner, I've really enjoyed tinkering with the codebase as I learn all aspects of MuZero.  I see in the MuZero paper they describe how they mask the policy logits to allowable moves in the root node of the MCTS:

> AlphaZero used the set of legal actions obtained from the simulator to mask the prior
produced by the network everywhere in the search tree. MuZero only masks legal actions at the root of the search tree where the environment can be queried, but does not perform any masking within the search tree. This is possible because the network rapidly learns not to predict actions that never occur in the trajectories it is trained on.

Do you think the MCTS should mask out illegal moves if it is at the root node (if the game being played supports it)?  It might speed up the learning process for these types of games.  If so, do you want me to send you a pull request for it?

**Comment by werner-duvaud** [comment]
Hi,
You are right, we are aware of that. We developed this feature but it was not useful for gym games.

We are working on a ResNet for board games and we will surely add the possibility to mask the actions at the root at the same time, but we want this to remain an option to stay general. I find it also elegant because, without masking illegal actions, MuZero really learns without prior knowledge of the game.

As you said, this forces the game to calculate the moves allowed each time. For some games it takes much longer to calculate all possible moves in advance rather than verifying the legality of just one.

An alternative is to consider an illegal move as the end of the game, this forces MuZero to learn early on which moves are legal or not, just like in the rest of the MCTS.

The version with the ResNet will leave the possibility of masking or not the actions at the root. We will push it soon, thank you for the proposal.

**Comment by SunilaAkbar** [comment]
Hi Werner,

Thanks so much for sharing the code.

I am using it for my task scheduling problem. I find it's not converging for my problem, although I tried number of times by playing around with the hyperparameters. I was thinking that the reason could be no provision of the legal actions in MCTS search. I just wanted to ask that whether you have the version with the provision of legal actions in MCTS search. If yes, please do share, I will be very grateful.

Thanks and Best Regards,
Sunila
